gramps/plugins/tool/reorderids.py

```diff
@@ -34,207 +35,516 @@
+                # (implication logical boolean operator below)
+                if act_id.startswith(prefix) or not keep_fmt:
+                    obj.set_gramps_id(new_id)
+                    commit(obj, self.trans)
+                    new_id = self.obj_values[prim_obj].succ_id()
+            else:
+                # attempt to extract integer - if we can't, treat it as a duplicate
+                try:
+                    match = _findint.match(act_id)
+                    if match:
+                        # get the integer, build the new handle. Make sure it
+                        # hasn't already been chosen. If it has, put this
+                        # in the duplicate handle list
+
+                        index = int(match.groups()[0])
+                        if formatmatch:
+                            if index > index_max:
+                                new_id = next_from_id()
+                            else:
+                                new_id = prefix_fmt % index
                         else:
-                            newgramps_id = prefix % int(index)
-                    else:
-                        # the prefix does not contain a number after %, eg I%d
-                        newgramps_id = prefix % int(index)
+                            # prefix_fmt does not contain a number after %, eg I%d
+                            new_id = prefix_fmt % index
 
-                    if newgramps_id == gramps_id:
-                        if newgramps_id in newids:
-                            dups.append(obj.get_handle())
+                        if new_id == act_id:
+                            if new_id in new_ids:
+                                dup_ids.append(obj.get_handle())
+                            else:
+                                new_ids[new_id] = act_id
+                        elif get_from_id(new_id) is not None:
+                            dup_ids.append(obj.get_handle())
                         else:
-                            newids[newgramps_id] = gramps_id
-                    elif find_from_id(newgramps_id) is not None:
-                        dups.append(obj.get_handle())
+                            obj.set_id(new_id)
+                            commit(obj, self.trans)
+                            new_ids[new_id] = act_id
                     else:
-                        obj.set_gramps_id(newgramps_id)
-                        commit(obj, self.trans)
-                        newids[newgramps_id] = gramps_id
-                else:
-                    dups.append(handle)
-            except:
-                dups.append(handle)
+                        dup_ids.append(handle)
+                except:
+                    dup_ids.append(handle)
 
         # go through the duplicates, looking for the first available
         # handle that matches the new scheme.
-
-        if self.uistate:
-            self.progress.set_pass(_('Finding and assigning unused IDs'),
```

Comment:
Have you tried it on a large family tree 50k+  that's why the progress meter is used?